コメント行をインデント対象に含めるオプションを追加した。

coeracanth opened the pull request:
settings.jsonに`"eraindent.indentCommentRow": true`と記載するとコメント行の自動インデントが有効になります。
デフォルトは無効です。
```settings.json
{
    "files.autoGuessEncoding": true,
    "eraindent.indentCommentRow": true //←こんなかんじ
}
```
コメントの文脈は考慮されず、コード内のインデントに従います。
```erabasic
;xxxの場合
IF COND("xxx")
	PRINTFORML xxx
	;yyyの場合    ←こういうの
ELSEIF COND("yyy")
	PRINTFORML yyy
	;その他    ←こういうの
ELSE
	PRINTFORML zzz
ENDIF
```
```erabasic
IF COND("xxx")
	PRINTFORML xxx
	;DEBUGPRINT hoge ←こういうの
ELSEIF COND("yyy")
	PRINTFORML yyy
        ;○/○ 更新        ←こういうの　　　
        ;PRINTFORML aaa  ←こういうの
ELSE
	PRINTFORML zzz
ENDIF
```

他人のリポジトリにプルリクを出すのも、vscode拡張のソースをいじるのも初めてで勝手がわかっていないので、変なところありましたらご指摘お願いします。

Commit:
コメント行をインデント対象に含めるオプションを追加した。

lunany188 commented on src/indenter.ts:
```diff
@@ -606,10 +620,27 @@ export const getNextNewLine = (state: IndentState): [string, Line[]] | null => {
+    this._config = config;
+  }
+
+  private _config;
```
1/4
とりあえずまず一番の問題としてはココです。変数の型が明示されていないのが問題です。
現在コンパイルの設定として変数の型がanyと推定された場合、意図しないところで何が代入されるか管理することが難しく正しく取り扱うことが難しいため、コンパイルエラーとして弾くように設定してあります。anyがわからない場合はTypescriptの型の振る舞いについて追加で説明を加えますので質問してかまいません。
またスタイルとしてはインターフェイスにしろクラス変数にしろ関数の引数にしろ関数内のローカルな変数にしろ、変数を使う際はまずできる限りconstやreadonlyを用いて定数あるいは一度代入された値が変わらないように努め、どうしても(ループを通して値を計算したいなど)変数が必要な場合は型を明記してください。もちろんプライベートだろうが。

coeracanth replied:
2/4 と合わせて、IndenterConfig を削除しました。

lunany188 commented on src/indenter.ts:
```diff
@@ -606,10 +620,27 @@ export const getNextNewLine = (state: IndentState): [string, Line[]] | null => {
+export interface EraIndenterConfig {
+  get<T>(section: string, defaultValue: T): T;
+}
+export class IndenterConfig {
```
2/4
またそもそもとしてこのclass自体が必要ないかもしれません。
EraIndenterConfigというインターフェイス自体はVSCodeのConfig取得との境界面なので必要ですがそれからの値の取り出し方をクラスにまとめなくてもいいと思います。

coeracanth replied:
1/4 と合わせて、IndenterConfig を削除しました。

lunany188 commented on src/indenter.ts:
```diff
@@ -141,15 +141,17 @@ export function isEraIndenterError(object: any): object is EraIndenterError {
+  constructor(option: EraIndentorOptions, config: EraIndenterConfig) {
     this.state = makeIndentState(option);
     this.previous = this.state;
+    this.config = new IndenterConfig(config);
```
3/4
そして外から受け入れたConfigについては以前別にお伝えした通りIndentStateの中にまとめて詰め込んでしまってください。…がちょっとここは設計がめんどくさいのでいっそこのままでも受け入れられます。
もしできるのであればIndentStateに、IndenterConfigとしてではなくその中からindentCommentRowを取り出したうえで、入れるようにmakeIndentStateとupdateIndentStateを書き換えてください。makeIndentStateの引数はEraIndenterConfigのままでいいのでmakeIndentStateに対してEraIndenterConfigから必要な値を取り出す責務を負わせるようにするといいと思います。
どうせここら辺今見直すと汚い処理してるのであとで手直しする予定ですから先ほど言った通りこのままでも受け入れられます。

coeracanth replied:
IndenterConfig の中身を IndentState に統合しました。

updateIndentState の実装についてですが、s に u をマージする目的なら Object.assign() が使えるかと思いますがどうでしょうか。
https://github.com/lunany188/vscode-eraIndent/blob/73fd347086b4e90b3470dcb6c5948f1d7c7f3d69/src/indenter.ts#L81-L99
https://github.com/lunany188/vscode-eraIndent/blob/208130cf03d739873c8dc1b8322da3c148c70179/src/indenter.ts#L88-L93

lunany188 replied:
些細なところですがせっかくTypescriptを使っているので`Object.assign`を使うよりもspread syntaxを使うのがより良いと思います。
``` typescript
 export const updateIndentState = ( 
   s: IndentState, 
   u: Partial<IndentState> 
 ): IndentState => { 
   return {...s, ...u}; 
 }; 
```
すでにあるオブジェクトを別のオブジェクトで上書きするのであれば`Object.assign`のほうが意図としてより適切でしょうが、今回はオブジェクトを合成した新しいオブジェクトを作りたいのでこちらの方が`{}`を作って引数に積み上げる処理を挟まず直接新しく生成したオブジェクトに書き込む分わずかに高速になります。
新しい構文なのでブラウザ対応状況が悪いのが問題ですがそれはトランスパイルの条件として動作環境を適切に設定すれば自動で`Object.assign`を使うようにtypescriptがうまいこと変換してくれるはずです。(まあVSCodeだとそのままspread構文として展開されるでしょう)
とはいえそれの修正はこちらの手元でもできるのでこれで受け入れたいと思います。

lunany188 commented on src/indenter.ts:
```diff
@@ -332,14 +340,17 @@ export function getNewIndent(
 export function getNewIndentNormal(
-  lineState: LineState
+  lineState: LineState,
+  config: IndenterConfig
```
4/4
そして修正を加える場合ここの引数がIndenterConfigであることはよくないですね。
stateをまるまる引き渡すか必要な内容だけ取り出して引き渡すかこの関数で必要な変数を引き渡すためだけの新しいインターフェイスを作るか…まあ雑でいいと思います。
個人のスタイルの好みとしてはgetNewIndentのstateの型表示を参考にしてみてください。このように引数に与えられたオブジェクトからその関数内で必要なプロパティだけを明示してとりだすよううな書き方が書いた当時の自分にとって好ましかったようです…現在の自分はさすがに冗長じゃないかと思ったりもしますがまあ実際に作るならそれでもこんな感じに手癖で描いてしまうことでしょう。

coeracanth replied:
IndenterConfig の中身を IndentState に統合し、IndentState を引数に渡すよう変更しました。

lunany188 commented:
とりあえず大まかに動作させるために最低限必要な修正の1点を含めたスタイルの話を4点ほど指摘事項を記載しました。よろしくお願いします。
ロジックのほつれは特段存在せず(というかむしろ自分が昔書いたコードのほうにそれをいくつか見つけました…)最低限指摘した型の明記さえ行われたならば受け入れることができるでしょう。クラスの削除やらIndenterState周りについてはどうせ別個に手を入れるので。
実際にバージョンを上げる前にはもう少し手元で別に作業が必要ですが先行して進めると衝突の解決がめんどくさいですしこれを受け入れてから進めて来月中にはアップデートを出せればというところです。

lunany188 commented:
あー少し読み間違えがあったので少し指摘事項修正しますね…書いてて気づいた

lunany188 commented:
これでよろしくお願いします

lunany188 reviewed (changes requested):
よろしくお願いします

Commit:
指摘事項を受けての修正

- class IndenterConfig を削除した。
- IndentState に IndenterConfig の内容を含めるように変更した。
- 関数 getNewIndentNormal の引数を IndenterConfig から IndentState に変更した。
